hopsangeneratorgui/src/hopsangeneratorgui.cpp

```diff
@@ -525,6 +530,91 @@ bool HopsanGeneratorGUI::checkComponentLibrary(const QString& libraryXMLPath)
     return checkOK;
 }
 
+
+bool HopsanGeneratorGUI::addComponentToLibrary(const QString &libraryXmlPath, const QString &cafPath)
+{
+    auto lw = mPrivates->createNewWidget();
+    loadGeneratorLibrary();
+
+    constexpr auto functionName = "callAddExistingComponentToLibrary";
+    const auto libpath = libraryXmlPath.toStdString();
+    MessageForwarder forwarder(lw->widget());
+    const auto xmlpath = libraryXmlPath.toStdString();
+    const auto cafpath = cafPath.toStdString();
+
+    using AddExistingComponentToLibraryFunction_t = bool(const char*, const char*, MessageHandler_t, void*);
+    bool checkOK = mPrivates->call<AddExistingComponentToLibraryFunction_t>(forwarder, functionName, xmlpath.c_str(), cafpath.c_str(), &messageHandler, static_cast<void*>(&forwarder));
+    lw->setDidSucceed(checkOK);
+    return checkOK;
+}
+
+
+bool HopsanGeneratorGUI::addComponentToLibrary(const QString &libraryXmlPath, const QString &typeName, const QString &displayName, const QString &cqsType,
+                                               const QStringList &constantNames, const QStringList &constantDescriptions, const QStringList &constantUnits, const QStringList &constantInits,
+                                               const QStringList &inputNames, const QStringList &inputDescriptions, const QStringList &inputUnits, const QStringList &inputInits,
+                                               const QStringList &outputNames, const QStringList &outputDescriptions, const QStringList &outputUnits, const QStringList &outputInits,
+                                               const QStringList &portNames, const QStringList &portDescriptions, const QStringList &portTypes, const QList<bool> &portsRequired)
+{
+    auto lw = mPrivates->createNewWidget();
+    loadGeneratorLibrary();
+
+    constexpr auto functionName = "callAddComponentToLibrary";
+    const auto libpath = libraryXmlPath.toStdString();
+    MessageForwarder forwarder(lw->widget());
+    const auto xmlpath = libraryXmlPath.toStdString();
+    const auto typenamestr = typeName.toStdString();
```

Comment:
subtype is missing, but I am not sure if it is needed in this case anyway, when you have written a new component. Also is should be merged into typename.... later

Reply:
You can easily work around this by first creating the component and then modifying it's CAF file to add the subtype. We can perhaps add this to the generator later.